tests/functional/test_ec2.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from tests import unittest
+from botocore.stub import Stubber, StubAssertionError
+import botocore.session
+
+
+class TestIdempotencyToken(unittest.TestCase):
+    def setUp(self):
+        self.region = 'us-west-2'
+        self.session = botocore.session.get_session()
+        self.client = self.session.create_client(
+            'ec2', self.region)
+        self.stubber = Stubber(self.client)
+
+    def test_insert_idempotency_token(self):
+        service_response = {}
+        expected_params_notoken = {
+            'PurchaseRequests': [
+                {'PurchaseToken': 'foo',
+                 'InstanceCount': 123}],
+        }
+        expected_params_token = {
+            'PurchaseRequests': [
+                {'PurchaseToken': 'foo',
+                 'InstanceCount': 123}],
+            'ClientToken': 'foobar'
+        }
+        function_name = 'purchase_scheduled_instances'
+
+        # First one should inject a token, second one is passed one
+        self.stubber.add_response(
+            function_name, service_response, expected_params_notoken)
+        self.stubber.add_response(
+            function_name, service_response, expected_params_token)
+
+        self.stubber.activate()
+
+        # Check the ClientToken parameter exsits
+        # after injection and before the call
+        self.client.meta.events.register_first(
```

Comment:
This is sort of confusing on what is happening here. For this functional test, here is what I recommend:


1) Split these tests out to two test methods: one where the user provides the parameter and one where you inject it for the user. In general if you are making two somewhat unrelated assertions, it is a sign you should break them up into separate tests because if the first assertion fails you will not be able to see if the second assertion fails as well.
2) For the case where you are injecting the idempotency token for the user, I would recommend instead of registering that lambda event, use ``botocore.stub.ANY``
3) For stubbers, feel free to use the context manager to activate and deactivate the stubber.